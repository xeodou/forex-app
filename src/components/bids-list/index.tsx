import { CurrencyRate } from "lib/forex-api/client";
import { FC } from "react";

export const BidsList: FC<{ rates: CurrencyRate[] }> = ({ rates }) => {
  return (
    <div className="bids-list w-full">
      <header className="flex justify-around uppercase">
        <span>CUR</span>
        <span>Time</span>
        <span>Ask</span>
        <span>Bid</span>
      </header>
      <div className="h-60 overflow-y-auto flex flex-col">
        {rates.map((rate, index) => (
          <ul
            key={rate.time_stamp}
            className={`flex justify-around ${index % 2 === 0 ? "bg-slate-300" : "bg-slate-100 "}`}
          >
            <li>{`${rate.from}/${rate.to}`}</li>
            <li>{new Date(rate.time_stamp).toLocaleString()}</li>
            <li>{rate.ask}</li>
            <li>{rate.bid}</li>
          </ul>
        ))}
      </div>
    </div>
  );
};
