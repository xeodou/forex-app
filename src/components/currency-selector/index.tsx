import { FC } from "react";
import { currencies } from "utils/currencies";

export interface CurrencyHeaderProps {
  // On selected currency changes
  onSelected?: (currency: string) => void;
  // Specific the default selected currency
  value?: string;
}

export const CurrencySelector: FC<CurrencyHeaderProps> = ({ onSelected, value }) => {
  const onChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    onSelected && onSelected(event.target.value);
  };

  return (
    <select
      onChange={onChange}
      value={value}
      className="h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm "
    >
      {currencies.map((currency) => (
        <option key={currency} value={currency}>
          {currency}
        </option>
      ))}
    </select>
  );
};
