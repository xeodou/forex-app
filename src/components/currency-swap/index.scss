.currency-header {
  &--indicator {
    width: 30px;
  }
}
