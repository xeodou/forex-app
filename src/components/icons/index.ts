export * from "./icon-swap";
export * from "./icon-refresh";
export * from './icon-close';
