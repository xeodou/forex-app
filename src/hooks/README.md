## Hooks

The shared hooks for the projects. Please read more if you want to know how does hooks work in React at [here](https://reactjs.org/docs/hooks-intro.html).

### use-currency-pair

This hook that will mantain a [CurrencyPairContext](./use-currency-pair/index.tsx?l=23) and shared the selected currency pair with the other component needs to access the currency pair.

Usage:

```typescript
import { useCurrencyPair } from "hooks";

const [currencyPair, setCurrencyPair] = useCurrencyPair();
```

### use-forex-rates

The hook wrap the forex getRates API into a hook, so you can read the realtime currency rates by React hook.

Usage:

```typescript
import { useForexRates } from "hooks";

const {
  // The realtime forex rates list
  forexRates,
  // The error response
  forexError,
  // A function to reset the data
  reset,
} = useForexRates();
```

### use-notify

This is a global hooks that will matain the notify message context.

Usage:

```typescript
import { useNotify } from "hooks";

const { stack, removeNotifyItem } = useNotify();
```
