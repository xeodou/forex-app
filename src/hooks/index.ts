export * from "./use-currency-pair";
export * from "./use-forex-rates";
