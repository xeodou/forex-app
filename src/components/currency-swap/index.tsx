import { CurrencySelector } from "components/currency-selector";
import { IconSwap } from "components/icons";
import { useCurrencyPair } from "hooks/use-currency-pair";

export const CurrencySwap = ({}) => {
  const [currencyPair, setCurrencyPair] = useCurrencyPair();

  const onFromCurrencyChange = (from: string) => {
    setCurrencyPair([from, currencyPair[1]]);
  };

  const onToCurrencyChange = (to: string) => {
    setCurrencyPair([currencyPair[0], to]);
  };

  const onSwitchClick = () => {
    setCurrencyPair([currencyPair[1], currencyPair[0]]);
  };

  return (
    <div className="flex items-center">
      <CurrencySelector onSelected={onFromCurrencyChange} value={currencyPair[0]} />
      <button
        type="button"
        aria-label="Switch currencies"
        className="font-medium text-sm p-2.5 text-center inline-flex items-center"
        onClick={onSwitchClick}
      >
        <IconSwap></IconSwap>
      </button>
      <CurrencySelector onSelected={onToCurrencyChange} value={currencyPair[1]} />
    </div>
  );
};
