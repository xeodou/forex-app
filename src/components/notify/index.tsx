import { useNotify } from "hooks";
import { Item } from "./_item";

export const Notify = () => {
  const { stack, removeNotifyItem } = useNotify();

  const lastItem = stack[stack.length - 1];

  const onDismiss = (id: number) => {
    removeNotifyItem(id);
  };

  return <div className="notify-container">{lastItem && <Item item={lastItem} onDismiss={onDismiss} />}</div>;
};
