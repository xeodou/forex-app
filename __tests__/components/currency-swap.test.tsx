import { act, fireEvent } from "@testing-library/react";
import { render, unmountComponentAtNode } from "react-dom";
import pretty from "pretty";
import { CurrencySwap } from "components/currency-swap";
import { CurrencyPairProvider, defaultCurrencyPair } from "hooks";

describe("currency-swap", () => {
  let container: HTMLDivElement;

  const Wrapper = () => (
    <CurrencyPairProvider initialValue={defaultCurrencyPair}>
      <CurrencySwap />
    </CurrencyPairProvider>
  );

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null as unknown as HTMLDivElement;
  });

  it("should render the selector", () => {
    act(() => {
      render(<Wrapper />, container);
    });

    expect(pretty(container.innerHTML)).toMatchSnapshot();

    expect(container.querySelectorAll("select")?.length).toBe(2);
    expect(container.querySelector("button")?.getAttribute("aria-label")).toEqual("Switch currencies");
  });

  it("should switch the currency pair", () => {
    act(() => {
      render(<Wrapper />, container);
    });

    expect(container.querySelectorAll("select")?.length).toBe(2);
    const button = container.querySelector("button") as HTMLButtonElement;
    expect(button).toBeTruthy();

    fireEvent.click(button);

    expect(pretty(container.querySelectorAll("select")[0].value)).toEqual(defaultCurrencyPair[1]);
    expect(pretty(container.querySelectorAll("select")[1].value)).toEqual(defaultCurrencyPair[0]);

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("should change the from currency", () => {
    act(() => {
      render(<Wrapper />, container);
    });

    expect(container.querySelectorAll("select")?.length).toBe(2);
    const from = container.querySelectorAll("select")[0] as HTMLSelectElement;
    expect(from).toBeTruthy();

    fireEvent.change(from, { target: { value: "CNY" } });

    expect(pretty(container.querySelectorAll("select")[0].value)).toEqual("CNY");

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("should change the to currency", () => {
    act(() => {
      render(<Wrapper />, container);
    });

    expect(container.querySelectorAll("select")?.length).toBe(2);
    const to = container.querySelectorAll("select")[1] as HTMLSelectElement;
    expect(to).toBeTruthy();

    fireEvent.change(to, { target: { value: "CNY" } });

    expect(pretty(container.querySelectorAll("select")[1].value)).toEqual("CNY");

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});
