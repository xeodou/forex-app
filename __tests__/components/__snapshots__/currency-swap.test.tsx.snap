// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`currency-swap should change the from currency 1`] = `
"<div class=\\"flex items-center\\"><select class=\\"h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm \\">
    <option value=\\"USD\\">USD</option>
    <option value=\\"JPY\\">JPY</option>
    <option value=\\"BGN\\">BGN</option>
    <option value=\\"CNY\\">CNY</option>
  </select><button type=\\"button\\" aria-label=\\"Switch currencies\\" class=\\"font-medium text-sm p-2.5 text-center inline-flex items-center\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"w-6 h-6\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" stroke-width=\\"1\\">
      <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" d=\\"M8 7h12m0 0l-4-4m4 4l-4 4m0 6H4m0 0l4 4m-4-4l4-4\\"></path>
    </svg></button><select class=\\"h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm \\">
    <option value=\\"USD\\">USD</option>
    <option value=\\"JPY\\">JPY</option>
    <option value=\\"BGN\\">BGN</option>
    <option value=\\"CNY\\">CNY</option>
  </select></div>"
`;

exports[`currency-swap should change the to currency 1`] = `
"<div class=\\"flex items-center\\"><select class=\\"h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm \\">
    <option value=\\"USD\\">USD</option>
    <option value=\\"JPY\\">JPY</option>
    <option value=\\"BGN\\">BGN</option>
    <option value=\\"CNY\\">CNY</option>
  </select><button type=\\"button\\" aria-label=\\"Switch currencies\\" class=\\"font-medium text-sm p-2.5 text-center inline-flex items-center\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"w-6 h-6\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" stroke-width=\\"1\\">
      <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" d=\\"M8 7h12m0 0l-4-4m4 4l-4 4m0 6H4m0 0l4 4m-4-4l4-4\\"></path>
    </svg></button><select class=\\"h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm \\">
    <option value=\\"USD\\">USD</option>
    <option value=\\"JPY\\">JPY</option>
    <option value=\\"BGN\\">BGN</option>
    <option value=\\"CNY\\">CNY</option>
  </select></div>"
`;

exports[`currency-swap should render the selector 1`] = `
"<div class=\\"flex items-center\\"><select class=\\"h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm \\">
    <option value=\\"USD\\">USD</option>
    <option value=\\"JPY\\">JPY</option>
    <option value=\\"BGN\\">BGN</option>
    <option value=\\"CNY\\">CNY</option>
  </select><button type=\\"button\\" aria-label=\\"Switch currencies\\" class=\\"font-medium text-sm p-2.5 text-center inline-flex items-center\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"w-6 h-6\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" stroke-width=\\"1\\">
      <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" d=\\"M8 7h12m0 0l-4-4m4 4l-4 4m0 6H4m0 0l4 4m-4-4l4-4\\"></path>
    </svg></button><select class=\\"h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm \\">
    <option value=\\"USD\\">USD</option>
    <option value=\\"JPY\\">JPY</option>
    <option value=\\"BGN\\">BGN</option>
    <option value=\\"CNY\\">CNY</option>
  </select></div>"
`;

exports[`currency-swap should switch the currency pair 1`] = `
"<div class=\\"flex items-center\\"><select class=\\"h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm \\">
    <option value=\\"USD\\">USD</option>
    <option value=\\"JPY\\">JPY</option>
    <option value=\\"BGN\\">BGN</option>
    <option value=\\"CNY\\">CNY</option>
  </select><button type=\\"button\\" aria-label=\\"Switch currencies\\" class=\\"font-medium text-sm p-2.5 text-center inline-flex items-center\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"w-6 h-6\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" stroke-width=\\"1\\">
      <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" d=\\"M8 7h12m0 0l-4-4m4 4l-4 4m0 6H4m0 0l4 4m-4-4l4-4\\"></path>
    </svg></button><select class=\\"h-full py-0 pl-4 pr-7 border-transparent bg-transparent text-gray-500 sm:text-sm \\">
    <option value=\\"USD\\">USD</option>
    <option value=\\"JPY\\">JPY</option>
    <option value=\\"BGN\\">BGN</option>
    <option value=\\"CNY\\">CNY</option>
  </select></div>"
`;
